src/OSPSuite.Core/Domain/Mappers/NeighborhoodCollectionToContainerMapper.cs

```diff
@@ -21,40 +24,72 @@ internal class NeighborhoodCollectionToContainerMapper : INeighborhoodCollection
             .WithMode(ContainerMode.Logical)
             .WithName(Constants.NEIGHBORHOODS);
 
-         var startValuesForFloatingMolecules = presentMoleculesCachedByContainerPath(moleculeNames, simulationBuilder);
+         var allSpatialStructureAndMergeBehaviors = simulationBuilder.SpatialStructureAndMergeBehaviors;
+         if (!allSpatialStructureAndMergeBehaviors.Any())
+            return neighborhoodsParentContainer;
 
-         var moleculeNamesCopyProperties = simulationBuilder.AllPresentXenobioticFloatingMoleculeNames();
+         var mapToNeighborhood = mapToNeighborhoodDef(modelConfiguration);
+
+         IReadOnlyList<Neighborhood> mapNeighborhoods(SpatialStructure spatialStructure) =>
+            spatialStructure.Neighborhoods.Select(mapToNeighborhood).Where(x => x != null).ToList();
 
          //we use a cache to ensure that we are replacing neighborhoods defined in multiple structures
-         var neighborhoodCache = new ObjectBaseCache<Neighborhood>();
-         var allNeighborhoodBuilder = simulationBuilder.SpatialStructureAndMergeBehaviors.SelectMany(x => x.spatialStructure.Neighborhoods);
-         neighborhoodCache.AddRange(allNeighborhoodBuilder.Select(nb =>
-               _neighborhoodMapper.MapFrom(nb, moleculeNamesFor(nb, startValuesForFloatingMolecules), moleculeNamesCopyProperties, modelConfiguration))
-            //can be null if the neighbors are not defined in the merged spatial structure
-            .Where(x => x != null));
-
-         neighborhoodsParentContainer.AddChildren(neighborhoodCache);
+         var firstSpatialStructure = allSpatialStructureAndMergeBehaviors[0].spatialStructure;
+         var allOtherSpatialStructuresWithMergeBehavior = allSpatialStructureAndMergeBehaviors.Skip(1).ToList();
+
+
+         //first step: Add the neighborhoods from the first structure
+         neighborhoodsParentContainer.AddChildren(mapNeighborhoods(firstSpatialStructure));
+
+         //now merge all other neighborhoods
+         allOtherSpatialStructuresWithMergeBehavior
+            .Select(x => new {x.mergeBehavior, neighborhoods = mapNeighborhoods(x.spatialStructure)})
+            .Each(x => mergeNeighborhoodsInStructure(neighborhoodsParentContainer, x.neighborhoods, x.mergeBehavior));
+
          return neighborhoodsParentContainer;
       }
 
+      private void mergeNeighborhoodsInStructure(IContainer neighborhoods, IReadOnlyList<Neighborhood> neighborhoodsToMerge, MergeBehavior mergeBehavior)
+      {
+         neighborhoodsToMerge.Each(neighborhoodToMerge =>
+         {
+            if (mergeBehavior == MergeBehavior.Extend)
+               _containerMergeTask.AddOrMergeContainer(neighborhoods, neighborhoodToMerge);
+            else
+               _containerMergeTask.AddOrReplaceInContainer(neighborhoods, neighborhoodToMerge);
+         });
+      }
+
+      private Func<NeighborhoodBuilder, Neighborhood> mapToNeighborhoodDef(ModelConfiguration modelConfiguration)
```

Comment:
This is SUCH a good pattern to avoid passing parameters around. Syntax is a big cumbersome but this allows for captured closure to be used where they need to and not necessarily at the top of a function